sentry/src/main/java/io/sentry/instrumentation/file/SentryFileInputStream.java

```diff
@@ -0,0 +1,144 @@
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * An implementation of {@link java.io.FileInputStream} that creates a {@link io.sentry.ISpan} for
+ * reading operation with filename and byte count set as description
+ *
+ * <p>Note, that span is started when this InputStream is instantiated via constructor and finishes
+ * when the {@link java.io.FileInputStream#close()} is called.
+ */
+@Open
+public class SentryFileInputStream extends FileInputStream {
+
+  private final @NotNull FileInputStream delegate;
+  private final @NotNull FileIOSpanManager spanManager;
+
+  public SentryFileInputStream(final @Nullable String name) throws FileNotFoundException {
+    this(name != null ? new File(name) : null, HubAdapter.getInstance());
+  }
+
+  public SentryFileInputStream(final @Nullable File file) throws FileNotFoundException {
+    this(file, HubAdapter.getInstance());
+  }
+
+  public SentryFileInputStream(final @NotNull FileDescriptor fdObj) {
+    this(fdObj, HubAdapter.getInstance());
+  }
+
+  public SentryFileInputStream(final @Nullable File file, final @NotNull IHub hub)
+      throws FileNotFoundException {
+    this(init(file, null, hub));
+  }
+
+  public SentryFileInputStream(final @NotNull FileDescriptor fdObj, final @NotNull IHub hub) {
+    this(init(fdObj, null, hub), fdObj);
+  }
+
+  private SentryFileInputStream(
+      final @NotNull FileInputStreamInitData data, final @NotNull FileDescriptor fd) {
+    super(fd);
+    spanManager = new FileIOSpanManager(data.span, data.file, data.hub);
+    delegate = data.delegate;
+  }
+
+  private SentryFileInputStream(final @NotNull FileInputStreamInitData data)
+      throws FileNotFoundException {
+    super(data.file);
+    spanManager = new FileIOSpanManager(data.span, data.file, data.hub);
+    delegate = data.delegate;
+  }
+
+  private static FileInputStreamInitData init(
+      final @Nullable File file, @Nullable FileInputStream delegate, final @NotNull IHub hub)
+      throws FileNotFoundException {
+    final ISpan span = FileIOSpanManager.startSpan(hub, "file.read");
+    if (delegate == null) {
+      delegate = new FileInputStream(file);
+    }
+    return new FileInputStreamInitData(file, span, delegate, hub);
+  }
+
+  private static FileInputStreamInitData init(
+      final @NotNull FileDescriptor fd,
+      @Nullable FileInputStream delegate,
+      final @NotNull IHub hub) {
+    final ISpan span = FileIOSpanManager.startSpan(hub, "file.read");
+    if (delegate == null) {
+      delegate = new FileInputStream(fd);
+    }
+    return new FileInputStreamInitData(null, span, delegate, hub);
+  }
+
+  @Override
+  public int read() throws IOException {
+    // this is the only case, when the read() operation returns the byte value, and not the count
+    // hence we need this special handling
+    AtomicInteger result = new AtomicInteger(0);
+    spanManager.performIO(
+        () -> {
+          result.set(delegate.read());
+          return result.get() != -1 ? 1 : 0;
```

Comment:
```suggestion
	        final read = delegate.read();
          result.set(read);
          return read != -1 ? 1 : 0;
```

its probably cheaper since AtomicInteger needs to be atomic